src/Interpreters/Streaming/Aggregator.cpp

```diff
@@ -103,6 +107,73 @@ Columns materializeKeyColumns(Columns & columns, ColumnRawPtrs & key_columns, co
 
     return materialized_columns;
 }
+
+Arena * getArena(AggregatedDataVariants & variants, AggregateStateType type)
+{
+    if (type == AggregateStateType::OnlyRetracted)
+        return variants.retracted_pool.get();
+    else
+        return variants.aggregates_pool;
+}
+
+template <typename BucketConverter>
+BlocksList concurrentBucketConvert(ThreadPool * thread_pool, const std::vector<Int64> & buckets, Arena * arena, Arenas & pools, BucketConverter && bucket_converter)
+{
+    std::atomic<UInt32> next_bucket_idx_to_merge = 0;
+    auto converter = [&](Arena * pool, ThreadGroupStatusPtr thread_group, const std::atomic_flag * cancelled) {
+        SCOPE_EXIT_SAFE(
```

Comment:
We can refactor out line 124 ~ 130 to `scheduleOrThrowOnError` lambda param at line 164 like 

```
auto converter = [&](Arena * pool, const std::atomic_flag * cancelled) {
        BlocksList blocks;
        while (true)
        {
            if (cancelled && cancelled->test())
                break;

            UInt32 bucket_idx = next_bucket_idx_to_merge.fetch_add(1);
            if (bucket_idx >= buckets.size())
                break;

            auto bucket = buckets[bucket_idx];
            blocks.splice(blocks.end(), bucket_converter(bucket, pool));
        }
        return blocks;
};
```


```
    for (size_t thread_id = 0; thread_id < num_threads; ++thread_id) 
    {
            thread_pool->scheduleOrThrowOnError([&pools, thread_id, group = CurrentThread::getGroup(), results, &converter, &cancelled] {
                CurrentThread::attachToIfDetached(group);
                SCOPE_EXIT_SAFE({CurrentThread::detachQueryIfNotDetached();});
                (*results)[thread_id] = converter(pools[thread_id].get(), &cancelled);
            });
    }
```

Reply:
applied it, but we need check whether the group is nullptr because `CurrentThread::getGroup()` maybe return nullptr